tfx/experimental/pipeline_testing/examples/chicago_taxi_pipeline/taxi_pipeline_regression_e2e_test.py

```diff
@@ -0,0 +1,152 @@
+    """
+    component_path = os.path.join(self._pipeline_root, component)
+    self.assertTrue(tf.io.gfile.exists(component_path))
+    outputs = tf.io.gfile.listdir(component_path)
+    for output in outputs:
+      execution = tf.io.gfile.listdir(os.path.join(component_path, output))
+      self.assertLen(execution, 1)
+
+  def testTaxiPipelineBeam(self):
+    record_dir = os.path.join(os.environ['HOME'],
+                              'tfx/tfx/experimental/pipeline_testing/',
+                              'examples/chicago_taxi_pipeline/testdata')
+
+    component_ids = ['CsvExampleGen', \
+                    'StatisticsGen', 'SchemaGen', \
+                    'ExampleValidator', 'Transform', \
+                    'Trainer', 'Evaluator', 'Pusher']
+
+    my_launcher = stub_component_launcher.create_stub_launcher_class(
+        record_dir,
+        component_ids,
+        {})
+    taxi_pipeline = taxi_pipeline_beam._create_pipeline(  # pylint:disable=protected-access, unexpected-keyword-arg
+                pipeline_name=self._pipeline_name,
+                data_root=self._data_root,
+                module_file=self._module_file,
+                serving_model_dir=self._serving_model_dir,
+                pipeline_root=self._pipeline_root,
+                metadata_path=self._metadata_path,
+                beam_pipeline_args=[])
+
+    BeamDagRunner(config=pipeline_config.PipelineConfig(
+        supported_launcher_classes=[
+            my_launcher,
+        ],
+        )).run(taxi_pipeline)
+
+    self.assertTrue(tf.io.gfile.exists(self._metadata_path))
+    metadata_config = metadata.sqlite_metadata_connection_config(
+        self._metadata_path)
+    with metadata.Metadata(metadata_config) as m:
+      artifact_count = len(m.store.get_artifacts())
+      execution_count = len(m.store.get_executions())
+      self.assertGreaterEqual(artifact_count, execution_count)
+      self.assertLen(taxi_pipeline.components, execution_count)
+
+    for component_id in component_ids:
+      self.assertExecutedOnce(component_id)
+
+    # Runs pipeline the second time.
+    BeamDagRunner(config=pipeline_config.PipelineConfig(
+        supported_launcher_classes=[
+            my_launcher,
+        ],
+        )).run(
+            taxi_pipeline_beam._create_pipeline(  # pylint:disable=protected-access, unexpected-keyword-arg
+                pipeline_name=self._pipeline_name,
+                data_root=self._data_root,
+                module_file=self._module_file,
+                serving_model_dir=self._serving_model_dir,
+                pipeline_root=self._pipeline_root,
+                metadata_path=self._metadata_path,
+                beam_pipeline_args=[]))
+
+    # All executions but Evaluator and Pusher are cached.
+    # Note that Resolver will always execute.
+    with metadata.Metadata(metadata_config) as m:
+      # Artifact count is increased by 3 caused by Evaluator and Pusher.
```

Comment:
Can you elaborate more on why this is 3? For example, `model` from pusher and `blessing` from evaluator.

And I cannot find anywhere the number 3 is used...